### PR TITLE
Mmca 3376

### PR DESCRIPTION
**Pull request by DamienButterworth**

### CODE REVIEW

File: app/models/file_upload/Nonce.scala [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+import scala.util.{Random, Try}
+
+/** Random integer value container. */
+sealed trait Nonce {
+  val value: Int
+}
+
+object Nonce {
+
+  final def random: Nonce =
```
Comment (derekspalmer) [inline app/models/file_upload/Nonce.scala:33]: Not sure this is particularly functional - breaks referential transparency.

File: app/models/file_upload/Nonce.scala [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    toNonce(value)
+
+  object Any extends Nonce {
+    final val value: Int = 0
+  }
+
+  final class Strict(val value: Int) extends Nonce {
+    override def equals(obj: scala.Any): Boolean =
+      obj match {
+        case _: Any.type => true
```
Comment (derekspalmer) [inline app/models/file_upload/Nonce.scala:46]: Isn't Any the super class of all types and hence won't this always be true?

File: app/models/file_upload/Nonce.scala [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+object Nonce {
+
+  final def random: Nonce =
+    toNonce(Random.nextInt())
+
+  final def apply(value: Int): Nonce =
+    toNonce(value)
+
+  object Any extends Nonce {
```
Comment (derekspalmer) [inline app/models/file_upload/Nonce.scala:39]: For Any to be a companion object needs to be declared in same file as class Any.

File: app/models/file_upload/Nonce.scala [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        case _ => false
+      }
+  }
+
+  implicit final val formats: Format[Nonce] =
+    SimpleDecimalFormat[Nonce](s => Nonce(s.toIntExact), n => BigDecimal(n.value))
+
+  implicit final def toNonce(value: Int): Nonce =
+    new Strict(value)
+}
```
Comment (derekspalmer) [inline app/models/file_upload/Nonce.scala:57]: Not sure I particularly understand this object.

File: test/models/NonceSpec.scala [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+package models
+
+import models.file_upload.Nonce
+import models.file_upload.Nonce.{Any, Strict}
+import utils.SpecBase
+
+class NonceSpec extends SpecBase {
+
+  "equals" should {
```
Comment (derekspalmer) [inline test/models/NonceSpec.scala:25]: Maybe it does work!

**Review by derekspalmer — APPROVED** [review]

**Review by derekspalmer — DISMISSED** [review]

**Review by zachhall1234 — APPROVED** [review]

**Pull request merged**